python-package/xgboost/sklearn.py

```diff
@@ -1153,7 +1172,9 @@ def _dmat_init(group, **params):
                               evals=evals,
                               evals_result=evals_result, feval=feval,
                               verbose_eval=verbose, xgb_model=xgb_model,
-                              callbacks=callbacks)
+                              callbacks=callbacks,
```

Comment:
Thanks for submitting a PR for this!  I like the idea of selecting range of iterations for evaluation.  But I really hope that we can implement it as a simple callback instead of parameters + callback.  Might need some refactoring.

Reply:
Thanks for the feedback! I think that's a good idea, I'll see what I can do

Reply:
@trivialfis Finally have some time to get back to this. 

I agree with you that it would be nice for this to be implemented as only a callback.  Its entirely possible that I'm missing something obvious but I'm not sure if it can be without making significant changes to the code. 

The main functionality added from this PR is the ability to skip the [bst.eval_set](https://github.com/dmlc/xgboost/blob/08dada08251e71ebd4e34a9ee788617f0a4c16e4/python-package/xgboost/training.py#L86-L93) method if the current iteration doesn't occur within the specified range (`score_tree_interval`). 

I suppose modifying `_train_internal` to  run [bst.eval_set](https://github.com/dmlc/xgboost/blob/08dada08251e71ebd4e34a9ee788617f0a4c16e4/python-package/xgboost/training.py#L86-L93) inside of a callback might work. This new score callback could be initialized with `eval_start` and `eval_interval` defaulting to 0 and 1 respectively. If the user wishes to change the defaults the user could manually add this callback with different initialization parameters. The only problem with this, would be that the state of `evaluation_result_list` would have to to be passed around between the different callbacks and the order that the callbacks are executed in would matter (this new callback would have to go first)

Happy to chat about this further, perhaps moving where the model evaluation occurs is the best solution

Reply:
@kylejn27  Let me take a deeper look.